### PR TITLE
GPL-787 Added check to labware barcodes for csv upload

### LINKED ISSUE #270: GPL-787 BUG - Uploading an incorrect csv to the 'Upload Labware' tab creates labware entity with same the barcode as an existing location
**Describe the bug**
By uploading an incorrect csv file it is possible to create labware entries with barcodes that match existing locations.

**RT Ticket Number**
N/A

**To Reproduce**
Steps to reproduce the behaviour:
1. Create a labware location for example _lw-tw-test-box-1-3701_
2. Create another labware location for example _lw-heron-boxes-destroyed-3700_
3. Upload a csv to http://uat.labwhere.psd.sanger.ac.uk/upload_labware/new of the form

```csv
Location,Labware,Position
lw-tw-test-box-1-3701,lw-heron-boxes-destroyed-3700,1
```

You will find a new piece of labware created under lw-tw-test-box-1-3701 with the same barcode as the existing location

**Expected behaviour**
I would expect the upload to fail and user to be informed of that the location column in their csv contains barcodes of locations.

**Screenshots**
Screenshot 2020-12-17 at 16 58 13

**Desktop (please complete the following information):**
Mac, Safari & Chrome

**Additional context**
This might explain some of the strange data in the reefer

### PR DESCRIPTION
**Pull request by BenTopping**
Closes #270 

Changes proposed in this pull request:

* Upload labwhere csv now has a check to make sure there are no labwares with the same barcode as an existing location
*

Comment: I believe the issue on #270  involving scan in/out having the same error as the one solved in this PR has previously been resolved

### CODE REVIEW

File: app/models/manifest_uploader.rb [hunk 1]
```diff
@@ -79,4 +84,10 @@ def check_for_missing_or_invalid_data
       end
     end
   end
+
+  def check_if_any_labwares_are_locations
+    if labwares.compact.any? { |labware| labware.match(/^#{Location::BARCODE_PREFIX}?/o) }
```
Comment (harrietc52) [inline app/models/manifest_uploader.rb:89]: "Upload labwhere csv now has a check to make sure there are no labwares with the same barcode as an existing location"
-> This code doesn't check that any Locations exist with the uploaded Labware barcode, but just that the uploaded Labware barcode doesn't start with `lw-`
Which is fine, just wanted to be clear. 
This is similar to the scan_form, which apparently wans't working? Does that need updating too? The difference is `.compact`, which I'm not sure how that makes the difference?!

**Review by harrietc52 — COMMENTED** [review]
minor comment

File: app/models/manifest_uploader.rb (hunk 1 above)
Comment (stevieing) [reply app/models/manifest_uploader.rb:89]: We have already added the fix to the scan form and it is in production. I was unsure on this story but occassionally they upload a dodgy csv which has location barcodes in the plate barcode column. The compact removes any nils as it is possible to scan nothing which would just add a line feed or an extra blank row. We don't want to throw an error as it is not user errror.

File: app/models/manifest_uploader.rb [hunk 2]
```diff
@@ -20,6 +21,10 @@ def formatted_data
     parsed.collect { |row| row.collect { |cell| cell.try(:strip) } }
   end
 
+  def labwares
+    @labwares ||= data.collect(&:last).uniq
```
Comment (harrietc52) [inline app/models/manifest_uploader.rb:25]: Can user .second instead

**Review by harrietc52 — CHANGES_REQUESTED** [review]
Comments we talked through :)

**Review by harrietc52 — DISMISSED** [review]

**Review by harrietc52 — APPROVED** [review]

**Pull request merged**